### PR TITLE
8271834: TestStringDeduplicationAgeThreshold intermittent failures on Shenandoah

### PR DESCRIPTION
**Pull request by zhengyu123**
To trigger young GC, the test fills heap with -Xmn amount of garbage. However, -Xmn does not apply to non-generational GC, such as Shenandoah, and -Xmn amount of garbage, sometimes, it not enough to trigger GC, therefore, there is no guarantee that strings can reach deduplication age threshold.

I purpose to use MXBean to get exact gc count to ensure expected gc cycles actually performed.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8271834](https://bugs.openjdk.java.net/browse/JDK-8271834): TestStringDeduplicationAgeThreshold intermittent failures on Shenandoah

### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**)
 * [Per Liden](https://openjdk.java.net/census#pliden) (@pliden - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5428/head:pull/5428` \
`$ git checkout pull/5428`

Update a local copy of the PR: \
`$ git checkout pull/5428` \
`$ git pull https://git.openjdk.java.net/jdk pull/5428/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5428`

View PR using the GUI difftool: \
`$ git pr show -t 5428`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5428.diff">https://git.openjdk.java.net/jdk/pull/5428.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/gc/stringdedup/TestStringDeduplicationTools.java [hunk 1]
```diff
@@ -89,21 +93,58 @@ private static void doFullGc(int numberOfTimes) {
         }
     }
 
+    private static volatile int gcCount;
+    private static NotificationListener listener = new NotificationListener() {
+        @Override
+        public void handleNotification(Notification n, Object o) {
+            if (n.getType().equals(GarbageCollectionNotificationInfo.GARBAGE_COLLECTION_NOTIFICATION)) {
+                GarbageCollectionNotificationInfo info = GarbageCollectionNotificationInfo.from((CompositeData) n.getUserData());
+                // Shenandoah GC also report GC pauses, skip them
```
Comment (shipilev) [inline test/hotspot/jtreg/gc/stringdedup/TestStringDeduplicationTools.java:102]:
```suggestion
                // Shenandoah GC also reports GC pauses, skip them
```

File: test/hotspot/jtreg/gc/stringdedup/TestStringDeduplicationTools.java [hunk 2]
```diff
@@ -89,21 +93,58 @@ private static void doFullGc(int numberOfTimes) {
...
+                if (info.getGcName().startsWith("Shenandoah")) {
+                    if ("end of GC cycle".equals(info.getGcAction())) {
+                        gcCount ++;
```
Comment (shipilev) [inline test/hotspot/jtreg/gc/stringdedup/TestStringDeduplicationTools.java:105]:
```suggestion
                        gcCount++;
```

File: test/hotspot/jtreg/gc/stringdedup/TestStringDeduplicationTools.java [hunk 3]
```diff
@@ -89,21 +93,58 @@ private static void doFullGc(int numberOfTimes) {
...
+                    }
+                } else {
+                    gcCount ++;
```
Comment (shipilev) [inline test/hotspot/jtreg/gc/stringdedup/TestStringDeduplicationTools.java:108]:
```suggestion
                    gcCount++;
```

**Review by shipilev — COMMENTED** [review]
It looks okay to me, but others need to have a look.

File: test/hotspot/jtreg/gc/stringdedup/TestStringDeduplicationTools.java (hunk 2 above)
Comment (zhengyu123) [reply test/hotspot/jtreg/gc/stringdedup/TestStringDeduplicationTools.java]: Fixed

File: test/hotspot/jtreg/gc/stringdedup/TestStringDeduplicationTools.java (hunk 3 above)
Comment (zhengyu123) [reply test/hotspot/jtreg/gc/stringdedup/TestStringDeduplicationTools.java]: Fixed

File: test/hotspot/jtreg/gc/stringdedup/TestStringDeduplicationTools.java (hunk 1 above)
Comment (zhengyu123) [reply test/hotspot/jtreg/gc/stringdedup/TestStringDeduplicationTools.java]: Fixed

File: test/hotspot/jtreg/gc/stringdedup/TestStringDeduplicationTools.java [hunk 4]
```diff
@@ -89,21 +93,58 @@ private static void doFullGc(int numberOfTimes) {
...
+    private static NotificationListener listener = new NotificationListener() {
+        @Override
+        public void handleNotification(Notification n, Object o) {
+            if (n.getType().equals(GarbageCollectionNotificationInfo.GARBAGE_COLLECTION_NOTIFICATION)) {
+                GarbageCollectionNotificationInfo info = GarbageCollectionNotificationInfo.from((CompositeData) n.getUserData());
+                // Shenandoah GC also reports GC pauses, skip them
+                if (info.getGcName().startsWith("Shenandoah")) {
+                    if ("end of GC cycle".equals(info.getGcAction())) {
+                        gcCount++;
+                    }
```
Comment (pliden) [inline test/hotspot/jtreg/gc/stringdedup/TestStringDeduplicationTools.java:106]: I think you want to do the same thing for ZGC here, which also reports both "pauses" and "cycles".

Comment (zhengyu123) [reply test/hotspot/jtreg/gc/stringdedup/TestStringDeduplicationTools.java]: Fixed. Thanks

**Review by pliden — APPROVED** [review]

**Review by shipilev — APPROVED** [review]

**Comment by zhengyu123** [comment]
Thanks, @shipilev @pliden

**Pull request closed without merging**